engines/cryomni3d/versailles/dialogs_manager.cpp

```diff
@@ -0,0 +1,367 @@
+	// Don't look for HNS file here
+	Common::String videoFName(_engine->prepareFileName(video, "hnm"));
+	Common::String soundFName(sound);
+
+	while (soundFName.size() < 8) {
+		soundFName += '_';
+	}
+	soundFName = _engine->prepareFileName(soundFName, "wav");
+
+	Video::HNMDecoder *videoDecoder = new Video::HNMDecoder(true);
+
+	if (!videoDecoder->loadFile(videoFName)) {
+		warning("Failed to open movie file %s/%s", video.c_str(), videoFName.c_str());
+		delete videoDecoder;
+		return;
+	}
+
+	Common::File *audioFile = new Common::File();
+	if (!audioFile->open(soundFName)) {
+		warning("Failed to open sound file %s/%s", sound.c_str(), soundFName.c_str());
+		delete videoDecoder;
+		delete audioFile;
+		return;
+	}
+
+	Audio::SeekableAudioStream *audioDecoder = Audio::makeWAVStream(audioFile, DisposeAfterUse::YES);
+	// We lost ownership of the audioFile just set it to nullptr and don't use it
+	audioFile = nullptr;
+
+	if (!audioDecoder) {
+		delete videoDecoder;
+		return;
+	}
+
+	g_system->showMouse(false);
+
+	uint16 width = videoDecoder->getWidth();
+	uint16 height = videoDecoder->getHeight();
+
+	// Preload first frame to draw subtitles from it
+	const Graphics::Surface *firstFrame = videoDecoder->decodeNextFrame();
+	assert(firstFrame != nullptr);
+
+	if (videoDecoder->hasDirtyPalette()) {
+		const byte *palette = videoDecoder->getPalette();
+		_engine->setupPalette(palette, 0, 256);
+	}
+
+	FontManager &fontManager = _engine->_fontManager;
+	_lastImage.create(firstFrame->w, firstFrame->h, firstFrame->format);
+	_lastImage.blitFrom(*firstFrame);
+
+	fontManager.setCurrentFont(7);
+	fontManager.setTransparentBackground(true);
+	fontManager.setForeColor(241);
+	fontManager.setLineHeight(22);
+	fontManager.setSpaceWidth(2);
+	fontManager.setCharSpacing(1);
+
+	if (_engine->showSubtitles()) {
+		Common::Rect block = settings.textRect;
+
+		unsigned int lines = fontManager.getLinesCount(text, block.width() - 8);
+		if (lines == 0) {
+			lines = 5;
+		}
+		unsigned int blockHeight = fontManager.lineHeight() * lines + 6;
+		block.setHeight(blockHeight);
+
+		if (block.bottom >= 480) {
+			block.bottom = 470;
+			warning("Dialog text is really too long");
+		}
+
+		// Make only the block area translucent inplace
+		Graphics::Surface blockSurface = _lastImage.getSubArea(block);
+		_engine->makeTranslucent(blockSurface, blockSurface);
+
+		fontManager.setSurface(&_lastImage);
+		block.grow(-4);
+		fontManager.setupBlock(block);
+		fontManager.displayBlockText(text);
+	}
+
+	g_system->copyRectToScreen(_lastImage.getPixels(), _lastImage.pitch, 0, 0, width, height);
+	g_system->updateScreen();
+
+	const Common::Rect &drawRect = settings.drawRect;
+
+	if (audioDecoder->getLength() == 0) {
+		// Empty wave file
+		delete audioDecoder;
+
+		unsigned int duration = 100 * text.size();
+		if (duration < 1000) {
+			duration = 1000;
+		}
+
+		bool skipWait = false;
+		unsigned int end = g_system->getMillis() + duration;
```

Comment:
If the duration is longer than 0.33 seconds here, the cursor lag could be noticeable. To avoid that, put `updateScreen()` and `delayMillis(10)` here (you've removed the delay from the `pollEvents()` earlier ;) )